File: collector/supervisor.py
"""Supervisor process: schedule and manage worker processes."""

import logging
import math
import signal
import sys
import time
from multiprocessing import Process

from .config import PRE_START_SECONDS, MARKET_INTERVAL, MARKET_DURATION, TAIL_SECONDS
from .worker import worker_main

log = logging.getLogger(__name__)


def _next_market_start(now: float) -> int:
    """Compute next eligible market_start such that we still have 70s lead time."""
    return math.ceil((now + PRE_START_SECONDS) / MARKET_INTERVAL) * MARKET_INTERVAL


def _reap_workers(
    active: dict[int, Process],
    restart_counts: dict[int, int],
    max_restarts: int = 1,
) -> None:
    """Join finished workers. Restart if died prematurely (once per market)."""
    finished = []
    for ms, proc in active.items():
        if not proc.is_alive():
            proc.join(timeout=1)
            finished.append(ms)

    now = time.time()
    for ms in finished:
        proc = active.pop(ms)
        expected_stop = ms + MARKET_DURATION + TAIL_SECONDS

        if now < expected_stop and proc.exitcode != 0:
            count = restart_counts.get(ms, 0)
            if count < max_restarts:
                log.warning(
                    "Worker %d died early (exit=%s), restarting (%d/%d)",
                    ms, proc.exitcode, count + 1, max_restarts,
                )
                p = Process(target=worker_main, args=(ms,), daemon=True)
                p.start()
                active[ms] = p
                restart_counts[ms] = count + 1
            else:
                log.error(
                    "Worker %d died early (exit=%s), max restarts reached",
                    ms, proc.exitcode,
                )
        else:
            log.info("Worker %d finished normally (exit=%s)", ms, proc.exitcode)


def main() -> None:
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s [supervisor] %(name)s %(levelname)s: %(message)s",
    )

    launched: set[int] = set()
    active: dict[int, Process] = {}
    restart_counts: dict[int, int] = {}

    # Graceful shutdown
    shutdown = False

    def _handle_signal(signum: int, frame: object) -> None:
        nonlocal shutdown
        log.info("Received signal %d, shutting down", signum)
        shutdown = True

    signal.signal(signal.SIGINT, _handle_signal)
    signal.signal(signal.SIGTERM, _handle_signal)

    log.info("Supervisor started, PID %d", __import__("os").getpid())

    while not shutdown:
        now = time.time()
        target = _next_market_start(now)
        launch_at = target - PRE_START_SECONDS

        sleep_secs = launch_at - time.time()
        if sleep_secs > 0:
            log.info(
                "Next market %d, launching at %d (sleeping %.1fs)",
                target, int(launch_at), sleep_secs,
            )
            # Sleep in small increments so we can respond to signals
            end = time.time() + sleep_secs
            while time.time() < end and not shutdown:
                time.sleep(min(1.0, end - time.time()))

        if shutdown:
            break

        if target not in launched:
            log.info("Spawning worker for market %d", target)
            p = Process(target=worker_main, args=(target,), daemon=True)
            p.start()
            launched.add(target)
            active[target] = p
            log.info("Worker for market %d started, PID %d", target, p.pid)

        _reap_workers(active, restart_counts)

        # Clean old entries from launched set (keep last hour)
        cutoff = now - 3600
        launched = {ms for ms in launched if ms > cutoff}

    # Shutdown: wait for active workers
    log.info("Shutting down, waiting for %d active workers", len(active))
    for ms, proc in active.items():
        log.info("Waiting for worker %d (PID %d)", ms, proc.pid)
        proc.join(timeout=30)
        if proc.is_alive():
            log.warning("Worker %d still alive, terminating", ms)
            proc.terminate()
            proc.join(timeout=5)

    log.info("Supervisor exited")

File: collector-rs/src/supervisor.rs
use std::collections::{HashMap, HashSet};
use std::time::{Duration, SystemTime, UNIX_EPOCH};

use tokio::task::JoinSet;
use tokio_util::sync::CancellationToken;

use bookview_collector::config::*;
use crate::worker::run_worker;

pub fn next_market_start(now_secs: f64) -> i64 {
    let adjusted = now_secs + PRE_START_SECONDS as f64;
    let boundary = (adjusted / MARKET_INTERVAL as f64).ceil() as i64;
    boundary * MARKET_INTERVAL as i64
}

fn now_secs() -> f64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs_f64()
}

pub async fn run(shutdown: CancellationToken) -> anyhow::Result<()> {
    let mut launched: HashSet<i64> = HashSet::new();
    let mut workers: JoinSet<(i64, Result<(), bookview_collector::error::WorkerError>)> = JoinSet::new();
    let mut restart_counts: HashMap<i64, u32> = HashMap::new();

    loop {
        let now = now_secs();
        let target = next_market_start(now);
        let launch_at = target as f64 - PRE_START_SECONDS as f64;
        let sleep_secs = launch_at - now_secs();

        if sleep_secs > 0.0 {
            tracing::info!(
                market = target,
                sleep_secs = format!("{:.1}", sleep_secs),
                "Next market, sleeping"
            );
            tokio::select! {
                _ = tokio::time::sleep(Duration::from_secs_f64(sleep_secs)) => {}
                _ = shutdown.cancelled() => break,
            }
        }

        if shutdown.is_cancelled() {
            break;
        }

        if !launched.contains(&target) {
            tracing::info!(market = target, "Spawning worker");
            let cancel = shutdown.child_token();
            let ms = target;
            workers.spawn(async move {
                let result = run_worker(ms, cancel).await;
                (ms, result)
            });
            launched.insert(target);
        }

        // Reap finished workers (non-blocking)
        while let Some(Ok((ms, result))) = workers.try_join_next() {
            match &result {
                Ok(()) => {
                    tracing::info!(market = ms, "Worker finished normally");
                }
                Err(e) => {
                    let now = now_secs();
                    let expected_stop = ms as f64 + (MARKET_DURATION + TAIL_SECONDS) as f64;
                    let count = restart_counts.entry(ms).or_insert(0);

                    if now < expected_stop && *count < MAX_RESTARTS {
                        *count += 1;
                        tracing::warn!(
                            market = ms,
                            error = %e,
                            attempt = *count,
                            "Worker died early, restarting"
                        );
                        let cancel = shutdown.child_token();
                        workers.spawn(async move {
                            let result = run_worker(ms, cancel).await;
                            (ms, result)
                        });
                    } else {
                        tracing::error!(
                            market = ms,
                            error = %e,
                            "Worker failed, max restarts reached"
                        );
                    }
                }
            }
        }

        // Prune launched set (keep last hour)
        let cutoff = (now_secs() - 3600.0) as i64;
        launched.retain(|&ms| ms > cutoff);
    }

    // Shutdown: cancel propagates to all child tokens automatically.
    // Wait for workers to finish.
    let active = workers.len();
    if active > 0 {
        tracing::info!(count = active, "Waiting for active workers to finish");
    }
    while let Some(result) = workers.join_next().await {
        match result {
            Ok((ms, Ok(()))) => {
                tracing::info!(market = ms, "Worker exited cleanly during shutdown");
            }
            Ok((ms, Err(e))) => {
                tracing::warn!(market = ms, error = %e, "Worker errored during shutdown");
            }
            Err(e) => {
                tracing::warn!(error = %e, "Worker task panicked during shutdown");
            }
        }
    }

    Ok(())
}
